### PR TITLE
[SPARK-35470][PYTHON] Fix disallow_untyped_defs mypy checks for pyspark.pandas.base

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Adds more type annotations in the file `python/pyspark/pandas/base.py` and fixes the mypy check failures.

### Why are the changes needed?

We should enable more disallow_untyped_defs mypy checks.

### Does this PR introduce _any_ user-facing change?

Yes.
This PR adds more type annotations in pandas APIs on Spark module, which can impact interaction with development tools for users.

### How was this patch tested?

The mypy check with a new configuration and existing tests should pass.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @HyukjinKwon @itholic @xinrong-databricks

File: python/pyspark/pandas/base.py [hunk 1]
```diff
@@ -58,11 +57,16 @@
 from pyspark.pandas.frame import DataFrame
 
 if TYPE_CHECKING:
+    from pyspark.pandas.data_type_ops.base import DataTypeOps  # noqa: F401 (SPARK-34943)
     from pyspark.pandas.indexes import Index  # noqa: F401 (SPARK-34943)
     from pyspark.pandas.series import Series  # noqa: F401 (SPARK-34943)
 
 
-def should_alignment_for_column_op(self: "IndexOpsMixin", other: "IndexOpsMixin") -> bool:
+T_IndexOps = TypeVar("T_IndexOps", bound="IndexOpsMixin")
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/base.py:65]: Is it a sort of convention? The naming seems a bit odd. I at least found one example in Python docs: https://docs.python.org/3/library/typing.html#typing.AnyStr

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

File: python/pyspark/pandas/base.py (hunk 1 above)
Comment (ueshin) [reply python/pyspark/pandas/base.py:65]: This is used for self-type annotation.
https://www.python.org/dev/peps/pep-0484/#annotating-instance-and-class-methods

```py
T = TypeVar('T', bound='Copyable')
class Copyable:
    def copy(self: T) -> T:
        # return a copy of self

class C(Copyable): ...
c = C()
c2 = c.copy()  # type here should be C
```

Comment (HyukjinKwon) [reply python/pyspark/pandas/base.py:65]: Oh no I mean the name of the class `T_IndexOps`

Comment (ueshin) [reply python/pyspark/pandas/base.py:65]: Ah, I see. I saw some similar examples in the built-in `typing.py`.

```py
T_co = TypeVar('T_co', covariant=True)  # Any type covariant containers.
V_co = TypeVar('V_co', covariant=True)  # Any type covariant containers.
VT_co = TypeVar('VT_co', covariant=True)  # Value type covariant containers.
T_contra = TypeVar('T_contra', contravariant=True)  # Ditto contravariant.
# Internal type variable used for Type[].
CT_co = TypeVar('CT_co', covariant=True, bound=type)
```

Comment (HyukjinKwon) [reply python/pyspark/pandas/base.py:65]: I see.. maybe we should better rename later in a separate PR ..

**Comment by HyukjinKwon** [comment]
Merged to master

File: python/pyspark/pandas/base.py (hunk 1 above)
Comment (ueshin) [reply python/pyspark/pandas/base.py:65]: Sure, let's track it in https://issues.apache.org/jira/browse/SPARK-35859.

**Pull request closed without merging**